### PR TITLE
meta module parse conditional imports

### PR DESCRIPTION
**Pull request by nickclmb**
Signed-off-by: Nikolay Samokhin <nickclmb@gmail.com>
Hi everyone,

At the moment if you have conditional imports in dart file, meta module will not have them in module dependencies:
```dart
import 'package:RAL/html_i.dart'
if (dart.library.html) 'dart:html'
if (dart.library.io) 'package:RAL/io/html_io.dart';
```
As a result the module will have dependency only on RAL/html_i.dart, missing RAL/io/html_io.dart
This pull request fixes that.

### CODE REVIEW

File: build_modules/lib/src/meta_module.dart [hunk 1]
```diff
@@ -89,10 +89,31 @@ class _AssetNode {
...
         }
         parts.add(linkedId);
-      } else if (internalSrcs.contains(linkedId)) {
-        internalDeps.add(linkedId);
-      } else {
-        externalDeps.add(linkedId);
+        continue;
+      }
+
+      List<Configuration> conditionalImportConfigurations;
```
Comment (jakemac53) [inline build_modules/lib/src/meta_module.dart]: nit: `conditionalDirectiveConfigurations`? technically this handles imports and exports.

**Comment by jakemac53** [comment]
Thanks for the pull request @nickclmb.

Eventually what we need here is actually separate modules per platform - where it would only include the config specific imports for that platform in the module. See https://github.com/dart-lang/build/issues/1430 for more context.

That being said - this would "work" for web projects I believe today, and might be ok as a stopgap solution. The main weird part is that DDC is going to end up compiling all the possible imports... not just the web specific ones.

If we can show with some tests that this works though (my guess is it will) - then I can see value in doing this.

@natebosch wdyt about doing this until we have proper support?

**Comment by jakemac53** [comment]
Fyi with the latest sdk I started getting this - reverting back my merge of the latest sdk for now:

```
===== CRASH =====
version=2.0.0-edge.f034f14b636affeaf4b6d26ea93aa5564b12eb25 (Wed Jun 6 09:05:13 2018 -0700) on "linux_x64"
si_signo=Segmentation fault(11), si_code=1, si_addr=0x120
Dumping native stack trace for thread 3bc25
  [0x000000000086d028] Unknown symbol
  [0x000000000086d028] Unknown symbol
  [0x0000000000aad982] dart::DRT_InlineCacheMissHandlerTwoArgs(dart::NativeArguments)
  [0x00007f69a574085b] Unknown symbol
  [0x00007f69a5741bde] Unknown symbol
  [0x00007f69a7113627] Unknown symbol
  [0x00007f69a424ee80] Unknown symbol
  [0x00007f69a424c8ce] Unknown symbol
  [0x00007f69a424c30b] Unknown symbol
  [0x00007f69a5740ce3] Unknown symbol
  [0x000000000090ef10] dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned long)
  [0x0000000000c3f70c] dart::Compiler::EvaluateStaticInitializer(dart::Field const&)
  [0x00000000009aba25] dart::Field::EvaluateInitializer() const
  [0x0000000000ab4002] dart::DRT_InitStaticField(dart::NativeArguments)
  [0x00007f69a574085b] Unknown symbol
  [0x00007f69a424c21a] Unknown symbol
  [0x00007f69a424bece] Unknown symbol
  [0x00007f69a4249c93] Unknown symbol
  [0x00007f69a4248ed7] Unknown symbol
  [0x00007f69a5740ce3] Unknown symbol
  [0x000000000090ef10] dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned long)
  [0x0000000000c7556c] Dart_Invoke
  [0x00000000006fe7a6] dart::bin::DartUtils::ResolveScript(_Dart_Handle*)
  [0x00000000006e4fb4] Unknown symbol
  [0x00000000006e3d78] Unknown symbol
  [0x00000000006e338b] dart::bin::RunMainIsolate(char const*, dart::bin::CommandLineOptions*)
  [0x00000000006e440d] dart::bin::main(int, char**)
  [0x00000000006e4c49] Unknown symbol
-- End of DumpStackTrace
```

File: build_modules/lib/src/meta_module.dart (hunk 1 above)
Comment (nickclmb) [reply build_modules/lib/src/meta_module.dart]: Got it, renamed

File: build_modules/lib/src/meta_module.dart [hunk 2]
```diff
@@ -89,10 +89,31 @@ class _AssetNode {
...
+      } else if (directive is ExportDirective && directive.configurations.isNotEmpty) {
+        conditionalImportConfigurations = directive.configurations;
+      }
+
+      final allDeps = <AssetId>[linkedId];
+      if (conditionalImportConfigurations != null) {
+        allDeps.addAll(conditionalImportConfigurations
+            .map((c) => Uri.parse(c.uri.value))
+            .where((u) => u.scheme != 'dart')
+            .map((u) => new AssetId.resolve(u.toString())));
```
[the remark this answers is not in the document]
Comment (nickclmb) [reply build_modules/lib/src/meta_module.dart]: Right, fixed that

**Comment by jakemac53** [comment]
Thanks @nickclmb, this LGTM.

It is worth also noting that given the current status as of [this issue](https://github.com/dart-lang/sdk/issues/32960) config specific imports will only support the platform constants (`dart.library.*`), at least for now.Once we add multi-platform support we will need to start enforcing that.

This will now technically be a breaking change - since this implementation is more flexible than that. I don't see that as a huge issue though just mentioning it for anybody who stumbles upon this PR and wants to start adding config specific imports to their code.

**Pull request merged**